sympy/solvers/ode/ode.py

```diff
@@ -1017,42 +1017,54 @@ class in it.  Note that a hint may do this anyway if
+        'nth_order_reducible': NthOrderReducible,
+        '2nd_nonlinear_autonomous_conserved': SecondNonlinearAutonomousConserved,
+        }
+
     ode = SingleODEProblem(eq_orig, func, x, prep=prep, xi=xi, eta=eta)
-    solvers = {
-        NthAlgebraic: ('nth_algebraic',),
-        FirstExact:('1st_exact',),
-        FirstLinear: ('1st_linear',),
-        AlmostLinear: ('almost_linear',),
-        Bernoulli: ('Bernoulli',),
-        Factorable: ('factorable',),
-        RiccatiSpecial: ('Riccati_special_minus2',),
-        SecondNonlinearAutonomousConserved: ('2nd_nonlinear_autonomous_conserved',),
-        Liouville: ('Liouville',),
-        Separable: ('separable',),
-        SeparableReduced: ('separable_reduced',),
-        HomogeneousCoeffSubsDepDivIndep: ('1st_homogeneous_coeff_subs_dep_div_indep',),
-        HomogeneousCoeffSubsIndepDivDep: ('1st_homogeneous_coeff_subs_indep_div_dep',),
-        HomogeneousCoeffBest: ('1st_homogeneous_coeff_best',),
-        LinearCoefficients: ('linear_coefficients',),
-        NthOrderReducible: ('nth_order_reducible',),
-        SecondHypergeometric: ('2nd_hypergeometric',),
-        NthLinearConstantCoeffHomogeneous: ('nth_linear_constant_coeff_homogeneous',),
-        NthLinearConstantCoeffVariationOfParameters: ('nth_linear_constant_coeff_variation_of_parameters',),
-        NthLinearConstantCoeffUndeterminedCoefficients: ('nth_linear_constant_coeff_undetermined_coefficients',),
-        NthLinearEulerEqHomogeneous: ('nth_linear_euler_eq_homogeneous',),
-        NthLinearEulerEqNonhomogeneousVariationOfParameters: ('nth_linear_euler_eq_nonhomogeneous_variation_of_parameters',),
-        NthLinearEulerEqNonhomogeneousUndeterminedCoefficients: ('nth_linear_euler_eq_nonhomogeneous_undetermined_coefficients',),
-        SecondLinearBessel: ('2nd_linear_bessel',),
-        SecondLinearAiry: ('2nd_linear_airy',),
-        LieGroup: ('lie_group',),
-    }
-    for solvercls in solvers:
-        solver = solvercls(ode)
+    user_hint = kwargs.get('hint','default')
+    if user_hint.endswith('_Integral'):
+        user_hint = user_hint[:-len('_Integral')]
+
+    if user_hint not in ['default', 'all', 'all_Integral', 'best'] and user_hint in solver_map:
+        solver = solver_map[user_hint](ode)
         if solver.matches():
-            for hints in solvers[solvercls]:
-                matching_hints[hints] = solver
-                if solvercls.has_integral:
-                    matching_hints[hints + "_Integral"] = solver
+            matching_hints[user_hint] = solver
+            if solver_map[user_hint].has_integral:
+                matching_hints[user_hint + "_Integral"] = solver
+    else:
+        for hint in solver_map:
+            solver = solver_map[hint](ode)
+            if solver.matches():
+                matching_hints[hint] = solver
+                if solver_map[hint].has_integral:
+                    matching_hints[hint + "_Integral"] = solver
```

Comment:
I would probably restructure this slightly:
```python
# Used when dsolve is called without an explicit hint.
# We exit early to return the first valid match
early_exit = (user_hint == 'default')

# An explicit hint has been given to dsolve
# Skip matching code for other hints
if user_hint not in ('default', 'all', 'all_Integral', 'best'):
    solver_map = ... # only the relevant items

for hint, cls in solver_map.items()
   ...
   if solver.matches():
       ...
       if early_exit:
           # maybe there needs to be a check here because of the unrefactored
           # solvers that are matched below. I guess this should only return if the
           # hint has a higher priority than all of those...
           return matching_hints
```
I suppose ultimately if all solvers were refactored then we wouldn't need to call `classify_ode` at all when a hint is given but for now the explicit hint case can be handled here.

Reply:
I am still trying to figure out what to do with the `default` hint as Factorable and NthorderReducible solvers call dsolve recursively.

Reply:
You have changed this code now but you seemed to have missed part of the point which is that by changing `solver_map` we can avoid duplicating the code that calls match.